src/librustc_mir/transform/clean_end_regions.rs

```diff
@@ -54,13 +56,32 @@ impl<'tcx> Visitor<'tcx> for GatherBorrowedRegions {
     fn visit_rvalue(&mut self,
                     rvalue: &Rvalue<'tcx>,
                     location: Location) {
+        // Gather regions that are used for borrows
         if let Rvalue::Ref(r, _, _) = *rvalue {
             if let RegionKind::ReScope(ce) = *r {
                 self.seen_regions.insert(ce);
             }
         }
         self.super_rvalue(rvalue, location);
     }
+
+    fn visit_statement(&mut self,
+                       block: BasicBlock,
+                       statement: &Statement<'tcx>,
+                       location: Location) {
+        self.super_statement(block, statement, location);
+    }
+
+    fn visit_ty(&mut self, ty: &Ty<'tcx>, _: Lookup) {
+        // Gather regions that occur in types
```

Comment:
No, we keep around all EndRegions for regions mentioned in any type anywhere. This still only concerns regions that are local to our function (otherwise, there wouldn't be an EndRegion) -- so, this covers all regions in the current function that are potentially interesting.

The way ordering works in the driver, `EraseEndRegions` runs before `AddValidate`, so we cannot just keep those that appear in validation statements.